content/blog/copilot-lessons/index.md

```diff
@@ -63,22 +61,24 @@ As Copilot grew, we broke it into smaller pieces we call skills. Each skill does
 
 When you ask Copilot something, it figures out what you need and picks the right skill for the job – like a manager deciding which expert to send your question to. This function-calling approach, orchestrated by a component we call the "outer loop," allows Copilot to access and process information beyond its internal knowledge base.
 
-To train this routing system, the team analyzed thousands of user queries and mapped them to Pulumi's underlying functionality. A question like 'Show me my untagged EC2 instances' breaks down into clear parameters - resource type (EC2), filter condition (untagged) - that route to the Insights skill. This mapping helped us handle the many ways users phrase the same technical request.
+Fortunately, Pulumi Cloud already exposes a [rich API](https://www.pulumi.com/docs/pulumi-cloud/reference/cloud-rest-api/) - in fact, this what powers the Pulumi Console and the Pulumi CLI - so all we had to do is to build a *skill* that maps the user query to the appropriate Pulumi Cloud REST API. A question like "Show my my stacks" translates into the [List Stacks](https://www.pulumi.com/docs/pulumi-cloud/reference/cloud-rest-api/#list-stacks) API call. A question like 'Show me my untagged EC2 instances' is a bit more complex but it breaks down into clear components - resource type (EC2), filter condition (untagged) - that route to the Insights skill. This mapping helped us handle the many ways users phrase the same technical request.
 
 Refining this routing system revealed another opportunity: streamlining the Debug button workflow.
 
 ## Debug Dispatch
 
 ![Before and After](optimize.png)
 
+<!-- We have an issue here with the picture: LLM does not call the API. LLM used to make up the URL (occasionally hallucinating), then Copilot would call that API and fed the results back the LLM for summarization. This is a very important point about how tool invocation work, we must get it right -->
```

Comment:
Issue with the picture, cc @adamgordonbell

Reply:
yeah, I'm changing it to Copilot. so Copilot calls the api, calls the LLM ect.